docs/learn/documentation/versioned/hadoop/overview.md

```diff
@@ -0,0 +1,46 @@
+   (the "License"); you may not use this file except in compliance with
+   the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+-->
+
+Samza provides a unified data processing model for both stream and batch processing. In Samza batch is treated as a special case of streaming: the sources are bounded data sets in batch, whereas unbounded data sets in streaming. Under the hood, it’s the same highly-efficient stream-processing engine.
+
+![diagram-medium](/img/{{site.version}}/learn/documentation/hadoop/unified_batch_streaming.png)
+
+### Unified API for Batch and Streaming
+
+Samza provides a single set of APIs for both batch and stream processing. This unified programming API makes it convenient for you to focus on the processing logic, without treating bounded source and unbounded source differently. Switching between batch and streaming only requires config change, e.g. [Kafka](../api/overview.html) to [HDFS](./consumer.html), instead of any code change.
+
+### Multi-stage Batch Pipeline
+
+Complex data pipelines usually consist multiple stages, with data shuffled (repartitioned) across stages for various windowing, aggregation and join. Samza [high-level API](/startup/preview/index.html) provides a useful operator named `partitionBy` to help the repartitioning of the data. Internally, Samza creates a physical stream, called an “intermediate stream”, sends the previous stage output to the stream, and then feed it to the next stage of the pipeline. The lifecycle of intermediate streams is completely managed by Samza so from the user perspective the data shuffling is automatic.
+
+For a single-stage pipeline, dealing with bounded data sets is straightforward: the system consumer “knows” the end of a particular partition, and it will emit end-of-stream token once a partition is complete. Samza will shut down the container when all its input partitions are complete.
+
+For a multi-stage pipeline, however, things become tricky since intermediate streams are often physically unbounded data streams, e.g. Kafka, and the downstream stages don't know when to shut down since unbounded streams don't have an end. To solve this problem, Samza uses in-band end-of-stream control messages in the intermediate stream along with user data messages. The upstream stage broadcasts end-of-stream control messages to every partition of the intermediate stream, and the downstream stage will aggregate the messages received and conclude an intermediate stream partition reaches to the end once the aggregation result matches all its upstream tasks. The following diagram shows the flow:
```

Comment:
...will aggregate the end-of-stream messages for each partition. When one end-of-stream message has been received for every upstream task in a partition, the downstream stage will conclude that the partition has no more messages...